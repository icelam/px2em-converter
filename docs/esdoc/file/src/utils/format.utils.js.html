<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <base data-ice="baseUrl" href="../../../">
  <title data-ice="title">src/utils/format.utils.js | PX2Em Converter</title>
  <link type="text/css" rel="stylesheet" href="css/style.css">
  <link type="text/css" rel="stylesheet" href="css/prettify-tomorrow.css">
  <script src="script/prettify/prettify.js"></script>
  <script src="script/manual.js"></script>
<meta name="description" content="PX2EM is a conversion tool which allows you to calculate the em sizes from pixels. Just add the base font size and the pixels to convert, and you&apos;re done!"><meta property="twitter:card" content="summary"><meta property="twitter:title" content="PX2Em Converter"><meta property="twitter:description" content="PX2EM is a conversion tool which allows you to calculate the em sizes from pixels. Just add the base font size and the pixels to convert, and you&apos;re done!"><meta property="twitter:image" content="./preview.png"></head>
<body class="layout-container" data-ice="rootContainer">

<header>
  <a href="./" style="display: flex; align-items: center;"><img src="./image/brand_logo.png" style="width:34px;"></a>
  
  <a href="identifiers.html">Reference</a>
  <a href="source.html">Source</a>
  <a href="test.html" data-ice="testLink">Test</a>
  <div class="search-box">
  <span>
    <img src="./image/search.png">
    <span class="search-input-edge"></span><input class="search-input"><span class="search-input-edge"></span>
  </span>
    <ul class="search-result"></ul>
  </div>
<a href="http://github.com/icelam/px2em-converter">Repository</a></header>

<nav class="navigation" data-ice="nav"><div>
  <ul>
    
  <li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#hooks">hooks</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-useDebounce">useDebounce</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#utils">utils</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-convertHexToRgb">convertHexToRgb</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-convertPixelRangeToArray">convertPixelRangeToArray</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-deduplicateArray">deduplicateArray</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-removeExtraSpacesAndCommas">removeExtraSpacesAndCommas</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-roundToNearest">roundToNearest</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-sortArrayAscending">sortArrayAscending</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-clearAll">clearAll</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getKey">getKey</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-removeKey">removeKey</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-setKey">setKey</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-isValidFontSize">isValidFontSize</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-isValidPositiveInteger">isValidPositiveInteger</a></span></span></li>
</ul>
</div>
</nav>

<div class="content" data-ice="content"><h1 data-ice="title">src/utils/format.utils.js</h1>
<pre class="source-code line-number raw-source-code"><code class="prettyprint linenums" data-ice="content">import { isValidPositiveInteger } from &apos;./validate.utils&apos;;

/**
 * Coverts any HEX color code to RGB
 * @param {number} value - Interger / floating point value that needs to be rounded
 * @param {number} precision - Number of decimal places
 * @returns {number} Rounded floating point value
 */
export const roundToNearest = (value, precision) =&gt; {
  if (!Number.isInteger(precision)) {
    return value;
  }
  // parseFloat for trimming trailing zeros
  return parseFloat(value.toFixed(precision));
};

/**
 * Deduplicates any array
 * @param {array} arr - Array to be deduplicated
 * @returns {array} Deduplicated array If `arr` provided is not an array, returns an empty array
 */
export const deduplicateArray = (arr) =&gt; {
  if (!Array.isArray(arr)) {
    return [];
  }

  const arraySet = new Set(arr);
  return Array.from(arraySet);
};

/**
 * Remove multiple consecutive spaces, trialing space and comma
 * @param {string} str - String to process
 * @returns {string} String which its multiple consecutive spaces,
 * trialing space and comma are removed
 */
export const removeExtraSpacesAndCommas = (str) =&gt; str
  .replace(/\s+/g, &apos; &apos;)
  .replace(/\s$/, &apos;&apos;)
  .replace(/,$/, &apos;&apos;);

/**
 * Sort array in ascending order
 * @param {array} arr - Array to sort
 * @returns {array} Array sorted in ascending order
 */
const sortArrayAscending = (arr) =&gt; {
  const newArray = [...arr];
  return newArray.sort((a, b) =&gt; a - b);
};

/**
 * Convert pixel range to list of pixels
 * @param {string} str - Pixel range, e.g. `&quot;12 - 20, 50, 100&quot;`
 * @returns {array} List of pixels, e.g. `[12, 13, 14, 15, 16, 17, 18, 19, 20, 50, 100]``
 */
export const convertPixelRangeToArray = (str) =&gt; {
  try {
    const pixels = [];

    if (!str) {
      throw new Error(&apos;EMPTY_INPUT&apos;);
    }

    const formattedString = removeExtraSpacesAndCommas(str);
    const rangeItems = formattedString.split(&apos;,&apos;);

    rangeItems.forEach((item) =&gt; {
      let currentDigit = &apos;&apos;;
      const operations = [];

      const pushDigit = () =&gt; {
        operations.push(currentDigit);
        currentDigit = &apos;&apos;;
      };

      /**
       * Parse string to operations:
       *  - &quot;11&quot; to [&quot;11&quot;]
       *  - &quot;11 - 13&quot; to [&quot;11&quot;, &quot;-&quot;, &quot;13&quot;]
       *  - &quot;1-3&quot; to [&quot;1&quot;, &quot;-&quot;, &quot;1&quot;]
       */
      for (let i = 0; i &lt; item.length; i++) {
        if (item[i] === &apos; &apos;) {
          currentDigit &amp;&amp; pushDigit();
        } else if (isValidPositiveInteger(item[i])) {
          currentDigit += item[i];
        } else if (item[i] === &apos;-&apos;) {
          currentDigit &amp;&amp; pushDigit();
          operations.push(&apos;-&apos;);
        } else {
          throw new Error(&apos;INVALID_INPUT&apos;);
        }
      }

      currentDigit &amp;&amp; pushDigit();

      /**
       * Convert operations to pixels:
       *  - Valid Case 1: Single positive integer
       *  - Valid Case 2: Pixel range (e.g. 1-10)
       */
      if (operations.length !== 1 &amp;&amp; operations.length !== 3) {
        throw new Error(&apos;INVALID_OPERATION_LENGTH&apos;);
      }

      if (operations.length === 1) {
        if (isValidPositiveInteger(operations[0])) {
          pixels.push(parseInt(operations[0], 10));
        } else {
          throw new Error(&apos;INVALID_NUMBER&apos;);
        }
      } else {
        const [leftString, operator, rightString] = operations;
        if (isValidPositiveInteger(leftString) &amp;&amp; operator === &apos;-&apos; &amp;&amp; isValidPositiveInteger(rightString)) {
          const leftNumber = parseInt(leftString, 10);
          const rightNumber = parseInt(rightString, 10);
          if (leftNumber &lt;= rightNumber) {
            for (let i = leftNumber; i &lt;= rightNumber; i++) {
              pixels.push(parseInt(i, 10));
            }
          } else {
            throw new Error(&apos;INVALID_RANGE&apos;);
          }
        } else {
          throw new Error(&apos;INVALID_OPERATION&apos;);
        }
      }
    });

    const uniquePixels = deduplicateArray(pixels);
    const sortedPixels = sortArrayAscending(uniquePixels);
    return sortedPixels;
  } catch (error) {
    return [];
  }
};
</code></pre>

</div>

<footer class="footer">
  Generated by <a href="https://esdoc.org">ESDoc<span data-ice="esdocVersion">(1.1.0)</span><img src="./image/esdoc-logo-mini-black.png"></a>
</footer>

<script src="script/search_index.js"></script>
<script src="script/search.js"></script>
<script src="script/pretty-print.js"></script>
<script src="script/inherited-summary.js"></script>
<script src="script/test-summary.js"></script>
<script src="script/inner-link.js"></script>
<script src="script/patch-for-local.js"></script>
</body>
</html>
