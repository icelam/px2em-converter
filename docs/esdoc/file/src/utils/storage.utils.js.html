<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <base data-ice="baseUrl" href="../../../">
  <title data-ice="title">src/utils/storage.utils.js | PX2EM Converter</title>
  <link type="text/css" rel="stylesheet" href="css/style.css">
  <link type="text/css" rel="stylesheet" href="css/prettify-tomorrow.css">
  <script src="script/prettify/prettify.js"></script>
  <script src="script/manual.js"></script>
<meta name="description" content="PX2EM is a conversion tool which allows you to calculate the em sizes from pixels. Just add the base font size and the pixels to convert, and you&apos;re done!"><meta property="twitter:card" content="summary"><meta property="twitter:title" content="PX2EM Converter"><meta property="twitter:description" content="PX2EM is a conversion tool which allows you to calculate the em sizes from pixels. Just add the base font size and the pixels to convert, and you&apos;re done!"><meta property="twitter:image" content="./preview.png"></head>
<body class="layout-container" data-ice="rootContainer">

<header>
  <a href="./" style="display: flex; align-items: center;"><img src="./image/brand_logo.png" style="width:34px;"></a>
  
  <a href="identifiers.html">Reference</a>
  <a href="source.html">Source</a>
  <a href="test.html" data-ice="testLink">Test</a>
  <div class="search-box">
  <span>
    <img src="./image/search.png">
    <span class="search-input-edge"></span><input class="search-input"><span class="search-input-edge"></span>
  </span>
    <ul class="search-result"></ul>
  </div>
<a href="http://github.com/icelam/px2em-converter">Repository</a></header>

<nav class="navigation" data-ice="nav"><div>
  <ul>
    
  <li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#hooks">hooks</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-useDebounce">useDebounce</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#utils">utils</a><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-convertHexToRgb">convertHexToRgb</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-convertPixelRangeToArray">convertPixelRangeToArray</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-deduplicateArray">deduplicateArray</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-removeExtraSpacesAndCommas">removeExtraSpacesAndCommas</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-roundToNearest">roundToNearest</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-sortArrayAscending">sortArrayAscending</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-clearAll">clearAll</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getKey">getKey</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-removeKey">removeKey</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-setKey">setKey</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-isValidFontSize">isValidFontSize</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-isValidPositiveInteger">isValidPositiveInteger</a></span></span></li>
</ul>
</div>
</nav>

<div class="content" data-ice="content"><h1 data-ice="title">src/utils/storage.utils.js</h1>
<pre class="source-code line-number raw-source-code"><code class="prettyprint linenums" data-ice="content">/**
 * Store data to local storage when key is provided
 * @param {string} key - Local storage key which specifies where the data to be stored
 * @param {*} data - Data to store to local storage
 */
const setKey = (key, data) =&gt; {
  if (!key) { return; }
  window.localStorage.setItem(key, data);
};

/**
 * Get the data stored in specific key of local storage
 * @property {function()} localStorageUtils - Helpers for interacting with local storage
 * @param {string} key - Local storage key which specifies where the data to be stored
 * @returns {string} Data stored in `key` of local storage
 */
const getKey = (key) =&gt; window.localStorage.getItem(key);

/**
 * Remove the data stored in specific key of local storage
 * @param {string} key - Local storage key which specifies where the data to be stored
 */
const removeKey = (key) =&gt; {
  window.localStorage.removeItem(key);
};

/**
 * Clear all data stored in local storage
 */
const clearAll = () =&gt; {
  window.localStorage.clear();
};

export default {
  setKey,
  getKey,
  removeKey,
  clearAll
};
</code></pre>

</div>

<footer class="footer">
  Generated by <a href="https://esdoc.org">ESDoc<span data-ice="esdocVersion">(1.1.0)</span><img src="./image/esdoc-logo-mini-black.png"></a>
</footer>

<script src="script/search_index.js"></script>
<script src="script/search.js"></script>
<script src="script/pretty-print.js"></script>
<script src="script/inherited-summary.js"></script>
<script src="script/test-summary.js"></script>
<script src="script/inner-link.js"></script>
<script src="script/patch-for-local.js"></script>
</body>
</html>
