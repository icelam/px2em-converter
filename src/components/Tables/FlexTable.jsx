import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const topRowBorderCss = css`
  border-width: 0.0625rem 0.0625rem 0.0625rem 0;
`;

const Table = styled.ul`
  &, & > li {
    border-style: solid;
    border-color: ${({ theme: { color } }) => color.borderColor};
  }

  & {
    display: flex;
    flex-flow: row wrap;
    list-style: none;
    padding: 0;
    margin: 0;
    border-width: 0 0 0 0.0625rem;

    > li {
      display: block;
      box-sizing: border-box;
      padding: 0.25rem 0.625rem;
      margin: 0;
      text-align: center;
      border-width: 0 0.0625rem 0.0625rem 0;
      flex: 0 1 ${(props) => 100 / props.sm}%;
      max-width: ${(props) => 100 / props.sm}%; /* IE 10 */

      &:nth-child(-n+${(props) => props.sm}){
        ${topRowBorderCss}
      }

      @media ${({ theme: { mediaQueries } }) => mediaQueries.tabletDesktop} {
        flex: 0 1 ${(props) => 100 / props.md}%;
        max-width: ${(props) => 100 / props.md}%;

        &:nth-child(-n+${(props) => props.md}){
          ${topRowBorderCss}
        }
      }

      @media ${({ theme: { mediaQueries } }) => mediaQueries.desktop} {
        flex: 0 1 ${(props) => 100 / props.lg}%;
        max-width: ${(props) => 100 / props.lg}%;

        &:nth-child(-n+${(props) => props.lg}){
          ${topRowBorderCss}
        }
      }

      @media ${({ theme: { mediaQueries } }) => mediaQueries.largerScreen} {
        flex: 0 1 ${(props) => 100 / props.xl}%;
        max-width: ${(props) => 100 / props.xl}%;

        &:nth-child(-n+${(props) => props.xl}){
          ${topRowBorderCss}
        }
      }
    }
  }
`;

const FlexTable = ({
  data, sm, md, lg, xl
}) => (
  <Table sm={sm} md={md} lg={lg} xl={xl} className="flex-table">
    {data.map(({ key, content }) => <li className="flex-table__cell" key={key}>{content}</li>)}
  </Table>
);

FlexTable.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string.isRequired,
    content: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired
  })).isRequired,
  sm: PropTypes.number,
  md: PropTypes.number,
  lg: PropTypes.number,
  xl: PropTypes.number
};

FlexTable.defaultProps = {
  sm: 1,
  md: 3,
  lg: 4,
  xl: 5
};

export default memo(FlexTable);
