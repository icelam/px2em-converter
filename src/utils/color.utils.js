const convertHexToRgb = (color) => {
  const colorCode = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(color.toLowerCase());

  return colorCode
    ? `${parseInt(colorCode[1], 16)}, ${parseInt(colorCode[2], 16)}, ${parseInt(colorCode[3], 16)}`
    : '';
};

export default convertHexToRgb;
