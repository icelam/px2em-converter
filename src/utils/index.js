export { default as convertHexToRgb } from './color.utils';
export * from './format.utils';
export { default as localStorageUtils } from './storage.utils';
export * from './validate.utils';
