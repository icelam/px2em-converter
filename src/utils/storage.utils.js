const setKey = (key, data) => {
  if (!key) { return; }
  window.localStorage.setItem(key, data);
};

const getKey = (key) => window.localStorage.getItem(key);

const removeKey = (key) => {
  window.localStorage.removeItem(key);
};

const clearAll = () => {
  window.localStorage.clear();
};

export default {
  setKey,
  getKey,
  removeKey,
  clearAll
};
