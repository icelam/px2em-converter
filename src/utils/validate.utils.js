/**
 * Validate if user input is a positive integer
 * @param {string} str - User input
 * @returns {boolean} `true` for positive integers, `false` for negative integers and other inputs
 */
export const isValidPositiveInteger = (str) => {
  const re = /^\d+$/;
  return re.test(str);
};

/**
 * Validate if user input is a valid font size
 * @param {string} str - User input
 * @returns {boolean} `true` for decimal and integer values, `false` for other inputs
 */
export const isValidFontSize = (str) => {
  const re = /^[0-9]+(\.[0-9]+)?$/;
  return re.test(str);
};
