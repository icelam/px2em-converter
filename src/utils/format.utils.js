import { isValidPositiveInteger } from './validate.utils';

export const roundToNearest = (value, precision) => {
  if (!Number.isInteger(precision)) {
    return value;
  }
  // parseFloat for trimming trailing zeros
  return parseFloat(value.toFixed(precision));
};

export const deduplicateArray = (arr) => {
  if (!Array.isArray(arr)) {
    return [];
  }

  const arraySet = new Set(arr);
  return Array.from(arraySet);
};

export const removeExtraSpacesAndCommas = (str) => str
  .replace(/\s+/g, ' ')
  .replace(/\s$/, '')
  .replace(/,$/, '');

const sortArrayAscending = (arr) => {
  const newArray = [...arr];
  return newArray.sort((a, b) => a - b);
};

export const convertPixelRangeToArray = (str) => {
  try {
    const pixels = [];

    if (!str) {
      throw new Error('EMPTY_INPUT');
    }

    const formattedString = removeExtraSpacesAndCommas(str);
    const rangeItems = formattedString.split(',');

    rangeItems.forEach((item) => {
      let currentDigit = '';
      const operations = [];

      const pushDigit = () => {
        operations.push(currentDigit);
        currentDigit = '';
      };

      /**
       * Parse string to operations:
       *  - "11" to ["11"]
       *  - "11 - 13" to ["11", "-", "13"]
       *  - "1-3" to ["1", "-", "1"]
       */
      for (let i = 0; i < item.length; i++) {
        if (item[i] === ' ') {
          currentDigit && pushDigit();
        } else if (isValidPositiveInteger(item[i])) {
          currentDigit += item[i];
        } else if (item[i] === '-') {
          currentDigit && pushDigit();
          operations.push('-');
        } else {
          throw new Error('INVALID_INPUT');
        }
      }

      currentDigit && pushDigit();

      /**
       * Convert operations to pixels:
       *  - Valid Case 1: Single positive integer
       *  - Valid Case 2: Pixel range (e.g. 1-10)
       */
      if (operations.length !== 1 && operations.length !== 3) {
        throw new Error('INVALID_OPERATION_LENGTH');
      }

      if (operations.length === 1) {
        if (isValidPositiveInteger(operations[0])) {
          pixels.push(parseInt(operations[0], 10));
        } else {
          throw new Error('INVALID_NUMBER');
        }
      } else {
        const [leftString, operator, rightString] = operations;
        if (isValidPositiveInteger(leftString) && operator === '-' && isValidPositiveInteger(rightString)) {
          const leftNumber = parseInt(leftString, 10);
          const rightNumber = parseInt(rightString, 10);
          if (leftNumber <= rightNumber) {
            for (let i = leftNumber; i <= rightNumber; i++) {
              pixels.push(parseInt(i, 10));
            }
          } else {
            throw new Error('INVALID_RANGE');
          }
        } else {
          throw new Error('INVALID_OPERATION');
        }
      }
    });

    const uniquePixels = deduplicateArray(pixels);
    const sortedPixels = sortArrayAscending(uniquePixels);
    return sortedPixels;
  } catch (error) {
    return [];
  }
};
