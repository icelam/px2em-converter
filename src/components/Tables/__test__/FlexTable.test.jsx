import React from 'react';
import { mediaQueries } from '@styles';
import { mountWithProvider } from '../../../../jest.utils';
import FlexTable from '../FlexTable';

const data = [
  {
    key: '1',
    content: <span>1</span>
  },
  {
    key: '2',
    content: <span>2</span>
  },
  {
    key: '3',
    content: <span>3</span>
  }
];

describe('FlexTable', () => {
  it('should render', () => {
    const wrapper = mountWithProvider(<FlexTable data={data} />);
    expect(wrapper.find('ul')).toHaveLength(1);
    expect(wrapper.find('ul li')).toHaveLength(3);
    expect(wrapper).toMatchSnapshot();
  });

  it('should display correct cell content', () => {
    const wrapper = mountWithProvider(<FlexTable data={data} />);
    expect(wrapper.find('ul li').at(0).find('span')).toHaveLength(1);
    expect(wrapper.find('ul li').at(0).text()).toEqual('1');
    expect(wrapper.find('ul li').at(1).find('span')).toHaveLength(1);
    expect(wrapper.find('ul li').at(1).text()).toEqual('2');
    expect(wrapper.find('ul li').at(2).find('span')).toHaveLength(1);
    expect(wrapper.find('ul li').at(2).text()).toEqual('3');
  });

  it('should adjust column width according to props', () => {
    const column = {
      sm: 1,
      md: 2,
      lg: 3,
      xl: 4
    };

    const wrapper = mountWithProvider(
      <FlexTable data={data} sm={column.sm} md={column.md} lg={column.lg} xl={column.xl} />
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'flex', `0 1 ${100 / column.sm}%`, { modifier: ' > li' }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'max-width', `${100 / column.sm}%`, { modifier: ' > li' }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'flex', `0 1 ${100 / column.md}%`, { modifier: ' > li', media: mediaQueries.tabletDesktop }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'max-width', `${100 / column.md}%`, { modifier: ' > li', media: mediaQueries.tabletDesktop }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'flex', `0 1 ${100 / column.lg}%`, { modifier: ' > li', media: mediaQueries.desktop }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'max-width', `${100 / column.lg}%`, { modifier: ' > li', media: mediaQueries.desktop }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'flex', `0 1 ${100 / column.xl}%`, { modifier: ' > li', media: mediaQueries.largerScreen }
    );
    expect(wrapper.find('ul').at(0)).toHaveStyleRule(
      'max-width', `${100 / column.xl}%`, { modifier: ' > li', media: mediaQueries.largerScreen }
    );
  });
});
