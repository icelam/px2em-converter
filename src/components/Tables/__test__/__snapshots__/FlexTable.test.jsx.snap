// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexTable should render 1`] = `
.c0,
.c0 > li {
  border-style: solid;
  border-color: #004b4d;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  list-style: none;
  padding: 0;
  margin: 0;
  border-width: 0 0 0 0.0625rem;
}

.c0 > li {
  display: block;
  box-sizing: border-box;
  padding: 0.25rem 0.625rem;
  margin: 0;
  text-align: center;
  border-width: 0 0.0625rem 0.0625rem 0;
  -webkit-flex: 0 1 100%;
  -ms-flex: 0 1 100%;
  flex: 0 1 100%;
  max-width: 100%;
}

.c0 > li:nth-child(-n+1) {
  border-width: 0.0625rem 0.0625rem 0.0625rem 0;
}

@media only screen and (min-width:576px) {
  .c0 > li {
    -webkit-flex: 0 1 33.333333333333336%;
    -ms-flex: 0 1 33.333333333333336%;
    flex: 0 1 33.333333333333336%;
    max-width: 33.333333333333336%;
  }

  .c0 > li:nth-child(-n+3) {
    border-width: 0.0625rem 0.0625rem 0.0625rem 0;
  }
}

@media only screen and (min-width:768px) {
  .c0 > li {
    -webkit-flex: 0 1 25%;
    -ms-flex: 0 1 25%;
    flex: 0 1 25%;
    max-width: 25%;
  }

  .c0 > li:nth-child(-n+4) {
    border-width: 0.0625rem 0.0625rem 0.0625rem 0;
  }
}

@media only screen and (min-width:992px) {
  .c0 > li {
    -webkit-flex: 0 1 20%;
    -ms-flex: 0 1 20%;
    flex: 0 1 20%;
    max-width: 20%;
  }

  .c0 > li:nth-child(-n+5) {
    border-width: 0.0625rem 0.0625rem 0.0625rem 0;
  }
}

<FlexTable
  data={
    Array [
      Object {
        "content": <span>
          1
        </span>,
        "key": "1",
      },
      Object {
        "content": <span>
          2
        </span>,
        "key": "2",
      },
      Object {
        "content": <span>
          3
        </span>,
        "key": "3",
      },
    ]
  }
>
  <styled.ul
    className="flex-table"
    lg={4}
    md={3}
    sm={1}
    xl={5}
  >
    <ul
      className="c0 flex-table"
    >
      <li
        className="flex-table__cell"
        key="1"
      >
        <span>
          1
        </span>
      </li>
      <li
        className="flex-table__cell"
        key="2"
      >
        <span>
          2
        </span>
      </li>
      <li
        className="flex-table__cell"
        key="3"
      >
        <span>
          3
        </span>
      </li>
    </ul>
  </styled.ul>
</FlexTable>
`;
